kraken-build/src/kraken/std/cargo/tasks/cargo_sync_config_task.py

```diff
@@ -36,10 +38,33 @@ def __init__(self, name: str, project: Project) -> None:
 
     def get_file_contents(self, file: Path) -> str | bytes:
         content = tomli.loads(file.read_text()) if not self.replace.get() and file.exists() else {}
-        content.setdefault("registry", {})["global-credential-providers"] = ["cargo:token"]
+        content.setdefault("registry", {})["global-credential-providers"] = [
+            "cargo:libsecret",
+            "cargo:macos-keychain",
+            "cargo:wincred",
+            "cargo:token",
+        ]
         content.setdefault("registries", {})["crates-io"] = {"protocol": self.crates_io_protocol.get()}
         for registry in self.registries.get():
+            publish_token = registry.publish_token
+            if publish_token is None:
+                continue
             content.setdefault("registries", {})[registry.alias] = {"index": registry.index}
+            p = run(
+                ["cargo", "login", "--registry", registry.alias],
+                cwd=self.project.directory,
+                capture_output=True,
+                input=publish_token.encode(),
+            )
+            if p.returncode != 0:
+                if p.stderr.endswith(b"\nerror: config.json not found in registry\n"):
+                    # expected error, ignore
+                    pass
+                else:
+                    # unknown error, fail normally
+                    stderr.write(p.stderr.decode())
+                    p.check_returncode()
+
```

Comment:
Agree on the above, just one question: Do we need to differentiate between build and publish tokens?

Reply:
Good point. It's even necessary: since `.cargo/config.toml` is not yet updated when still in the task, `cargo login` fails with `error: authenticated registries require a credential-provider to be configured`. I must have always had a previously-updated `.cargo/config.toml` in my tests.

I have added a task.